File: kpis/kpiLatency.m
function kpiLatency(obj, i, aebStartIdx)
    % AEB system latency
    %% Inputs:
    % obj            - KPIExtractor object
    % i              - Current index in the kpiTable
    % aebStartIdx    - Index of AEB request event

    % Use obj.kpiTable and obj.signalMatChunk directly
    kpiTable = obj.kpiTable;
    signalMatChunk = obj.signalMatChunk;
    cutoffFreq = obj.CUTOFF_FREQ;

    % Ensure required columns exist
    if ~ismember('m1IntvSysResp', kpiTable.Properties.VariableNames)
        obj.kpiTable.m1IntvSysResp = NaN(height(obj.kpiTable), 1);
    end
    if ~ismember('m1DeadTime', kpiTable.Properties.VariableNames)
        obj.kpiTable.m1DeadTime = NaN(height(obj.kpiTable), 1);
    end

    % Validate data range
    startIdx = aebStartIdx;
    endIdx = min(aebStartIdx + 30, length(signalMatChunk.time));
    if startIdx > endIdx
        warning('Invalid range for kneepoint detection at index %d. Setting m1IntvSysResp and m1DeadTime to NaN.', i);
        obj.kpiTable.m1IntvSysResp(i) = NaN;
        obj.kpiTable.m1DeadTime(i) = NaN;
        return;
    end

    % Detect kneepoint for AEB response
    [AEB_Resp, ~, ~] = utils.detectKneepoint_v4( ...
        signalMatChunk.time(startIdx:endIdx), ...
        signalMatChunk.longActAccel(startIdx:endIdx), ...
        'negative', cutoffFreq, 'curvature');
   
    % Calculate M0 and M1 with bounds check
    M0 = signalMatChunk.time(aebStartIdx);
    M1_idx = min(aebStartIdx + max(0, AEB_Resp), length(signalMatChunk.time) - 1);
    M1 = signalMatChunk.time(M1_idx);
    
    % Assign m1IntvSysResp with validation
    if isduration(M1)
        m1_value = seconds(M1);
    else
        m1_value = double(M1); % Handle numeric input
    end
    if ~isnan(m1_value) && isfinite(m1_value)
        obj.kpiTable.m1IntvSysResp(i) = round(m1_value, 2);
    else
        warning('Invalid M1 value for file index %d: %s. Setting m1IntvSysResp to NaN.', i, string(M1));
        obj.kpiTable.m1IntvSysResp(i) = NaN;
    end

    % Assign m1DeadTime with validation
    deadTime = M1 - M0;

    if isduration(deadTime)
        deadTime_value = seconds(deadTime);
    else
        deadTime_value = double(deadTime); % Handle numeric input
    end
    if ~isnan(deadTime_value) && isfinite(deadTime_value)
        obj.kpiTable.m1DeadTime(i) = round(deadTime_value, 2);
    else
        warning('Invalid deadTime value for file index %d: %s. Setting m1DeadTime to NaN.', i, string(deadTime));
        obj.kpiTable.m1DeadTime(i) = NaN;
    end

    % Display table row after assignment (convert to array first)
    rowData = table2array(obj.kpiTable(i, {'m1IntvSysResp', 'm1DeadTime'}));
end

File: python/kpis/aeb/kpi_latency.py
import numpy as np
import pandas as pd
import warnings
from utils.time_locators import detect_kneepoint_v4  # assuming you placed it there

def kpi_latency(obj, i, aeb_start_idx):
    """
    Compute AEB system latency KPIs.
    Updates obj.kpi_table in place.

    Parameters
    ----------
    obj : object with attributes
        - kpi_table : pandas.DataFrame
        - signal_mat_chunk : dict with arrays {time, longActAccel}
        - CUTOFF_FREQ : float
    i : int
        Row index in kpi_table.
    aeb_start_idx : int
        Index of AEB request event.
    """
    kpi_table = obj.kpi_table
    signals = obj.signal_mat_chunk
    cutoff_freq = obj.CUTOFF_FREQ

    # Ensure required columns exist
    for col in ["m1IntvSysResp", "m1DeadTime"]:
        if col not in kpi_table.columns:
            kpi_table[col] = pd.Series([np.nan] * len(kpi_table), dtype="float")

    # Validate data range
    start_idx = aeb_start_idx
    end_idx = min(aeb_start_idx + 30, len(signals["time"]) - 1)
    if start_idx > end_idx:
        warnings.warn(
            f"Invalid range for kneepoint detection at index {i}. "
            f"Setting m1IntvSysResp and m1DeadTime to NaN."
        )
        kpi_table.at[i, "m1IntvSysResp"] = np.nan
        kpi_table.at[i, "m1DeadTime"] = np.nan
        return

    # Detect kneepoint
    knee_idx, _, _ = detect_kneepoint_v4(
        time=signals["time"][start_idx:end_idx + 1],
        acc=signals["longActAccel"][start_idx:end_idx + 1],
        direction="negative",
        cutoff_freq=cutoff_freq,
        method="curvature",
    )

    # Compute M0 and M1
    M0 = signals["time"][aeb_start_idx]
    M1_idx = min(aeb_start_idx + max(0, knee_idx), len(signals["time"]) - 1)
    M1 = signals["time"][M1_idx]

    # Assign m1IntvSysResp
    try:
        m1_value = float(M1)  # assumes numeric time in seconds
    except Exception:
        m1_value = np.nan
    if np.isfinite(m1_value):
        kpi_table.at[i, "m1IntvSysResp"] = round(m1_value, 2)
    else:
        warnings.warn(f"Invalid M1 value for file index {i}: {M1}. Setting NaN.")
        kpi_table.at[i, "m1IntvSysResp"] = np.nan

    # Dead time (latency)
    try:
        dead_time_value = float(M1 - M0)
    except Exception:
        dead_time_value = np.nan
    if np.isfinite(dead_time_value):
        kpi_table.at[i, "m1DeadTime"] = round(dead_time_value, 2)
    else:
        warnings.warn(f"Invalid deadTime for file index {i}. Setting NaN.")
        kpi_table.at[i, "m1DeadTime"] = np.nan

    # Optional: return row as dict for debugging
    return kpi_table.loc[i, ["m1IntvSysResp", "m1DeadTime"]].to_dict()
